### PR TITLE
fix(AbstractIntervalDataFrameCursor): binary search scan

### PR DESCRIPTION
**Pull request by clickingbuttons**
In case of multiple equal timestamps, find the first when binary searching

### CODE REVIEW

File: core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java [hunk 1]
```diff
@@ -63,6 +63,16 @@ protected static long search(ReadOnlyColumn column, long value, long low, long h
                 low = mid + 1;
             else if (midVal > value)
                 high = mid;
+            else if (midVal > 0) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java:66]: we probably don't need to check that value is positive, timestamp can also be 0 and negative. Need to search for same values regardless

File: core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java [hunk 2]
```diff
@@ -63,6 +63,16 @@ protected static long search(ReadOnlyColumn column, long value, long low, long h
...
+                // In case of multiple equal values, find the first
+                int offset = 0;
+                long prevVal;
+                do {
+                    prevVal = column.getLong((mid + --offset) * 8);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java:71]: the getLong() arg can't be negative, there no check for it, which is what likely has caused CI to crash

File: core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java (hunk 1 above)
Comment (clickingbuttons) [reply core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java:66]: Ah sorry, I mean `mid > 0` to prevent ever doing `column.getLong(-1)`

Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java]: loop is walking backwards and can potentially step over the file boundary, so the test need to be in the loop itself

**Comment by clickingbuttons** [comment]
Any idea why this test would fail?
> Failed tests: 
  ConcurrentTest.testOneToParallelSubscriber:394 null

File: core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java (hunk 1 above)
Comment (clickingbuttons) [reply core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java]: Fixed.

File: core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java [hunk 3]
```diff
@@ -63,8 +63,15 @@ protected static long search(ReadOnlyColumn column, long value, long low, long h
                 low = mid + 1;
             else if (midVal > value)
                 high = mid;
-            else
+            else {
+                // In case of multiple equal values, find the first
+                int increment = scanUp ? -1 : 1;
+                for (long offset = 0; (mid + offset) > 0 && (mid + offset) < high; offset += increment) {
+                    if (midVal != column.getLong((mid + offset) * 8))
+                        return mid + offset - increment;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java:71]: logic is great, but
```
mid + offset
```
is calculated at least 3 times per loop. I hope Java would realise that and execute only once, but I would do this kind of optimisation manually

File: core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java [hunk 4]
```diff
@@ -63,8 +63,15 @@ protected static long search(ReadOnlyColumn column, long value, long low, long h
...
+                }
                 return mid;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java:73]: when loop goes all the way up with all values being the same the answer should be 0. Likewise if loop goes all the way down the answer should be end of column. `mid` is not a good outcome considering it is not mutated by the loop above.

Perhaps there is an opportunity to kill two birds with one stone and simplify loop by mutating `mid` ?

File: core/src/main/java/io/questdb/cairo/IntervalBwdDataFrameCursor.java [hunk 5]
```diff
@@ -76,12 +76,12 @@ public DataFrame next() {
 
                 // calculate intersection
 
-                long lo = search(column, intervalLo, 0, partitionLimit == -1 ? rowCount : partitionLimit);
+                long lo = search(column, intervalLo, 0, partitionLimit == -1 ? rowCount : partitionLimit, true);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/IntervalBwdDataFrameCursor.java:79]: may be here we can pass increment itself as constant, e.g.:

```java
static final int SCAN_UP = -1;
static final int SCAN_DOWN = 1;

...
long lo = search(column, intervalLo, 0, partitionLimit == -1 ? rowCount : partitionLimit, SCAN_UP);
```
this is both more readable and faster :)

File: core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java (hunk 3 above)
Comment (clickingbuttons) [reply core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java]: Fixed.

File: core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java (hunk 4 above)
Comment (clickingbuttons) [reply core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java:78]: Fixed.

File: core/src/main/java/io/questdb/cairo/IntervalBwdDataFrameCursor.java (hunk 5 above)
Comment (clickingbuttons) [reply core/src/main/java/io/questdb/cairo/IntervalBwdDataFrameCursor.java]: Fixed.

File: core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java (hunk 4 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java:78]: Perfect, thanks :+1:

**Pull request merged**